File: src/api/l2_ops.rs
// Copyright (c) 2026 qyiun666
// SPDX-License-Identifier: MIT OR Apache-2.0

//! API-5 + API-6: L2 Context CRUD and merge operations.

use crate::query::types::{
    ContextResult, MergeResult, SceneTreeResult, TopicDetail, TopicListQuery, TopicListResult,
    UpdateL2Fields,
};
use crate::shared::common::{format_hash, parse_id_to_hash};
use crate::{MemHop, MemHopError, Result};

impl MemHop {
    /// List L2 contexts with pagination and filtering.
    pub fn list_contexts(&self, scene_id: &str) -> Result<Vec<ContextResult>> {
        use crate::storage::record::REC_L2_TOPIC;
        let scene_hash = parse_id_to_hash(scene_id);
        let mut results = Vec::new();

        for (&id_hash, _) in self.engine.iter_index() {
            let Ok(Some((rt, data))) = self.engine.read_record(id_hash) else {
                continue;
            };
            if rt != REC_L2_TOPIC {
                continue;
            }
            let Ok(ctx) = bincode::deserialize::<crate::layers::context::ContextSlot>(data) else {
                continue;
            };
            if ctx.scene_id != scene_hash {
                continue;
            }
            results.push(slot_to_context_result(&ctx, id_hash, 0.0));
        }

        results.sort_by_key(|b| std::cmp::Reverse(b.user_timestamp));
        Ok(results)
    }

    /// Get a single L2 context by ID.
    pub fn get_context(&self, id: &str) -> Result<Option<ContextResult>> {
        use crate::storage::record::REC_L2_TOPIC;
        let id_hash = parse_id_to_hash(id);
        match self.engine.read_record(id_hash) {
            Ok(Some((rt, data))) if rt == REC_L2_TOPIC => {
                match bincode::deserialize::<crate::layers::context::ContextSlot>(data) {
                    Ok(ctx) => Ok(Some(slot_to_context_result(&ctx, id_hash, 0.0))),
                    Err(_) => Ok(None),
                }
            }
            _ => Ok(None),
        }
    }

    /// List the full scene tree for a given scene ID.
    pub fn list_context_tree(&self, scene_id: &str) -> Result<Vec<ContextResult>> {
        let scene_hash = parse_id_to_hash(scene_id);
        let tree = crate::query::l2_ops::list_scene_tree(&self.engine, &self.l2_meta, scene_hash)?;
        let results: Vec<ContextResult> = tree
            .nodes
            .into_iter()
            .map(|ctx| ContextResult {
                id: ctx.id.clone(),
                parent_id: ctx.parent_id.clone(),
                depth: ctx.depth,
                scene_id: ctx.scene_id.clone(),
                user_keywords: ctx.user_keywords.clone(),
                user_timestamp: ctx.user_timestamp,
                agent_keywords: ctx.agent_keywords.clone(),
                agent_timestamp: ctx.agent_timestamp,
                fused_keywords: ctx.fused_keywords.clone(),
                fused_summary: ctx.fused_summary.clone(),
                children_ids: ctx.children_ids.clone(),
                l4_refs: {
                    let mut refs = ctx.user_l4_refs.clone();
                    refs.extend(ctx.agent_l4_refs.clone());
                    refs.sort();
                    refs.dedup();
                    refs
                },
                l3_refs: {
                    let mut refs = ctx.user_l3_refs.clone();
                    refs.extend(ctx.agent_l3_refs.clone());
                    refs.sort();
                    refs.dedup();
                    refs
                },
                retrieval_score: 0.0,
            })
            .collect();
        Ok(results)
    }

    /// Partially update an L2 context and return ContextResult.
    pub fn update_context(&mut self, id: &str, fields: UpdateL2Fields) -> Result<ContextResult> {
        let detail =
            crate::query::l2_ops::update_l2(&mut self.engine, &mut self.sparse_index, id, fields)?;
        Ok(to_context_result(&detail, 0.0))
    }

    /// Delete an L2 context.
    pub fn delete_context(&mut self, id: &str) -> Result<()> {
        crate::query::l2_ops::delete_l2(
            &mut self.engine,
            &mut self.l1_reverse_index,
            &mut self.sparse_index,
            id,
        )
    }

    /// Merge multiple L2 contexts into one and return ContextResult.
    pub fn merge_contexts(&mut self, ids: Vec<String>) -> Result<ContextResult> {
        if ids.is_empty() {
            return Err(crate::MemHopError::InvalidQuery(
                "merge_contexts requires at least one ID".into(),
            ));
        }
        let primary_id = &ids[0];
        let merge_ids: Vec<String> = ids[1..].to_vec();
        let merge_result = crate::query::l2_ops::merge_l2(
            &mut self.engine,
            &mut self.sparse_index,
            primary_id,
            merge_ids,
        )?;
        // Fetch the merged primary context by ID and return as ContextResult
        self.get_context(&merge_result.primary_id)?.ok_or_else(|| {
            MemHopError::Serialization(format!(
                "Merged context {} not found after merge",
                merge_result.primary_id
            ))
        })
    }

    /// Delete multiple turns by their IDs.
    pub fn delete_turns(&mut self, ids: Vec<String>) -> Result<()> {
        for id in &ids {
            crate::query::l2_ops::delete_l2(
                &mut self.engine,
                &mut self.l1_reverse_index,
                &mut self.sparse_index,
                id,
            )?;
        }
        Ok(())
    }

    /// Legacy: list L2 contexts with pagination.
    pub fn list_l2(&self, query: TopicListQuery) -> Result<TopicListResult> {
        crate::query::l2_ops::list_l2(&self.engine, query)
    }

    /// Legacy: get a single L2 context by ID as TopicDetail.
    pub fn get_l2(&self, id: &str) -> Result<Option<TopicDetail>> {
        Ok(crate::query::l2_ops::get_l2(&self.engine, id)?
            .map(|ctx| crate::query::l2_ops::to_topic_detail(&ctx)))
    }

    /// Legacy: partially update an L2 context.
    pub fn update_l2(&mut self, id: &str, fields: UpdateL2Fields) -> Result<TopicDetail> {
        crate::query::l2_ops::update_l2(&mut self.engine, &mut self.sparse_index, id, fields)
    }

    /// Legacy: delete an L2 context.
    pub fn delete_l2(&mut self, id: &str) -> Result<()> {
        crate::query::l2_ops::delete_l2(
            &mut self.engine,
            &mut self.l1_reverse_index,
            &mut self.sparse_index,
            id,
        )
    }

    /// Legacy: delete a range of L4 archives from an L2 context.
    pub fn delete_turn(&mut self, id: &str, range: std::ops::Range<usize>) -> Result<()> {
        crate::query::l2_ops::delete_turn(&mut self.engine, &mut self.sparse_index, id, range)
    }

    /// Legacy: merge multiple L2 contexts.
    pub fn merge_l2(&mut self, primary_id: &str, merge_ids: Vec<String>) -> Result<MergeResult> {
        crate::query::l2_ops::merge_l2(
            &mut self.engine,
            &mut self.sparse_index,
            primary_id,
            merge_ids,
        )
    }

    /// Legacy: list the full scene tree.
    pub fn list_scene_tree(&self, scene_id: &str) -> Result<SceneTreeResult> {
        let scene_hash = parse_id_to_hash(scene_id);
        crate::query::l2_ops::list_scene_tree(&self.engine, &self.l2_meta, scene_hash)
    }
}

/// Convert a TopicDetail to a ContextResult
fn to_context_result(detail: &TopicDetail, score: f32) -> ContextResult {
    ContextResult {
        id: detail.id.clone(),
        parent_id: detail.parent_id.clone(),
        depth: detail.depth,
        scene_id: detail.scene_id.clone(),
        user_keywords: detail.user_keywords.clone(),
        user_timestamp: detail.user_timestamp,
        agent_keywords: detail.agent_keywords.clone(),
        agent_timestamp: detail.agent_timestamp,
        fused_keywords: detail.fused_keywords.clone(),
        fused_summary: detail.fused_summary.clone(),
        children_ids: detail.children_ids.clone(),
        l4_refs: {
            let mut refs = detail.user_l4_refs.clone();
            refs.extend(detail.agent_l4_refs.clone());
            refs.sort();
            refs.dedup();
            refs
        },
        l3_refs: {
            let mut refs = detail.user_l3_refs.clone();
            refs.extend(detail.agent_l3_refs.clone());
            refs.sort();
            refs.dedup();
            refs
        },
        retrieval_score: score,
    }
}

/// Convert a TopicDetail to a ContextResult using the TopicDetail directly (with id)
#[allow(dead_code)]
fn to_context_result_from_detail(detail: &TopicDetail, score: f32) -> ContextResult {
    to_context_result(detail, score)
}

/// Convert a ContextSlot to a ContextResult (pub(crate) for auto_create use)
use crate::layers::context::ContextSlot;
pub(crate) fn slot_to_context_result(ctx: &ContextSlot, id_hash: u64, score: f32) -> ContextResult {
    let id = format_hash(id_hash);
    ContextResult {
        id,
        parent_id: ctx.parent_id.map(format_hash),
        depth: ctx.depth,
        scene_id: format_hash(ctx.scene_id),
        user_keywords: ctx.user_keywords.clone(),
        user_timestamp: ctx.user_timestamp,
        agent_keywords: ctx.agent_keywords.clone(),
        agent_timestamp: ctx.agent_timestamp,
        fused_keywords: ctx.fused_keywords.clone(),
        fused_summary: ctx.fused_summary.clone(),
        children_ids: ctx.children_ids.iter().map(|&h| format_hash(h)).collect(),
        l4_refs: {
            let mut refs: Vec<String> = ctx
                .user_l4_refs
                .iter()
                .chain(ctx.agent_l4_refs.iter())
                .map(|h| format_hash(*h))
                .collect();
            refs.sort();
            refs.dedup();
            refs
        },
        l3_refs: {
            let mut refs: Vec<String> = ctx
                .user_l3_refs
                .iter()
                .chain(ctx.agent_l3_refs.iter())
                .map(|h| format_hash(*h))
                .collect();
            refs.sort();
            refs.dedup();
            refs
        },
        retrieval_score: score,
    }
}

File: memhop/internal/core/query/l2_ops.go
// Copyright (c) 2026 qyiun666
// SPDX-License-Identifier: MIT OR Apache-2.0

// L2 TopicSlot CRUD operations.

package query

import (
	"encoding/json"
	"sort"
	"strings"

	"github.com/qyiun666/memhop/memhop/internal/core"
	"github.com/qyiun666/memhop/memhop/internal/core/index"
	"github.com/qyiun666/memhop/memhop/internal/core/model"
	"github.com/qyiun666/memhop/memhop/internal/core/storage"
	"github.com/qyiun666/memhop/memhop/internal/hash"
	"github.com/qyiun666/memhop/memhop/internal/timeutil"
)

// GetL2 loads a single L2 context by hex ID.
func GetL2(engine *storage.StorageEngine, id string) (*model.TopicSlot, error) {
	idHash, err := hash.ParseID(id)
	if err != nil {
		return nil, core.NewError(core.ErrInvalidQuery, "parse l2 id", err)
	}
	return loadTopic(engine, idHash)
}

// ListL2 lists L2 contexts with pagination and keyword filter.
func ListL2(engine *storage.StorageEngine, q TopicListQuery) (*TopicListResult, error) {
	all := collectAllTopics(engine)
	filterByKeyword(&all, q.Keyword)
	sortTopicsByUpdated(all)
	skip, take := paginationParams(q.Page, q.PageSize)
	total := len(all)
	items := make([]TopicSummary, 0, take)
	for i := skip; i < skip+take && i < total; i++ {
		items = append(items, toTopicSummary(&all[i]))
	}
	return &TopicListResult{
		Items:    items,
		Total:    total,
		Page:     q.Page,
		PageSize: q.PageSize,
		HasMore:  skip+take < total,
	}, nil
}

// UpdateL2 partially updates an L2 context.
func UpdateL2(
	engine *storage.StorageEngine,
	sparse *index.SparseIndex,
	id string,
	fields UpdateL2Fields,
) (*TopicDetail, error) {
	idHash, err := hash.ParseID(id)
	if err != nil {
		return nil, core.NewError(core.ErrInvalidQuery, "parse l2 id", err)
	}
	ctx, err := loadTopic(engine, idHash)
	if err != nil {
		return nil, err
	}
	indexChanged := applyL2Updates(ctx, fields)
	if indexChanged {
		reindexTopic(sparse, ctx)
	}
	ctx.UpdatedAt = timeutil.NowMs()
	ctx.Version++
	if err := writeTopic(engine, idHash, ctx); err != nil {
		return nil, err
	}
	detail := toTopicDetail(ctx)
	return &detail, nil
}

// DeleteL2 deletes an L2 context and all associated data.
func DeleteL2(
	engine *storage.StorageEngine,
	l1Reverse *L1ReverseIndex,
	sparse *index.SparseIndex,
	id string,
) error {
	idHash, err := hash.ParseID(id)
	if err != nil {
		return core.NewError(core.ErrInvalidQuery, "parse l2 id", err)
	}
	ctx, err := loadTopic(engine, idHash)
	if err != nil {
		return nil // not found = already deleted
	}
	deleteL1Nodes(engine, l1Reverse, idHash)
	deleteL4Refs(engine, ctx)
	sparse.RemoveDocument(idHash)
	if l1Reverse != nil {
		l1Reverse.RemoveContext(idHash)
	}
	engine.DeleteRecord(idHash)
	return nil
}

// MergeL2 merges multiple L2 contexts into a primary context.
func MergeL2(
	engine *storage.StorageEngine,
	sparse *index.SparseIndex,
	primaryID string,
	mergeIDs []string,
) (*MergeResult, error) {
	primaryHash, err := hash.ParseID(primaryID)
	if err != nil {
		return nil, core.NewError(core.ErrInvalidQuery, "parse primary id", err)
	}
	if !engine.Contains(primaryHash) {
		return nil, core.NewError(core.ErrNotFound, "primary not found")
	}
	mergeHashes, err := parseMergeIDs(mergeIDs)
	if err != nil {
		return nil, err
	}
	primaryCtx, err := loadTopic(engine, primaryHash)
	if err != nil {
		return nil, err
	}
	absorbSecondaries(engine, sparse, primaryCtx, mergeHashes, mergeIDs)
	primaryCtx.UpdatedAt = timeutil.NowMs()
	primaryCtx.Version++
	if err := writeTopic(engine, primaryHash, primaryCtx); err != nil {
		return nil, err
	}
	reindexTopic(sparse, primaryCtx)
	turnCount := uint32(len(primaryCtx.UserL4Refs) + len(primaryCtx.AgentL4Refs))
	return &MergeResult{
		PrimaryID:        hash.FormatHash(primaryHash),
		MergedCount:      uint32(len(mergeHashes)),
		NewTurnCount:     turnCount,
		AbsorbedTopicIDs: mergeIDs,
	}, nil
}

// GetSceneTree lists the full tree of nodes within a scene.
func GetSceneTree(
	engine *storage.StorageEngine,
	l2Meta *index.L2MetaIndex,
	sceneID uint64,
) (*SceneTreeResult, error) {
	nodeIDs := l2Meta.GetByScene(sceneID)
	if len(nodeIDs) == 0 {
		return emptySceneTree(sceneID), nil
	}
	nodes := loadTopicsByIDs(engine, nodeIDs)
	sortTopicsByCreated(nodes)
	return buildSceneTree(sceneID, nodes), nil
}

// --- internal helpers ---

func loadTopic(engine *storage.StorageEngine, idHash uint64) (*model.TopicSlot, error) {
	_, data, err := engine.ReadRecord(idHash)
	if err != nil {
		return nil, err
	}
	var ctx model.TopicSlot
	if err := json.Unmarshal(data, &ctx); err != nil {
		return nil, core.NewError(core.ErrDeserialization, "unmarshal topic", err)
	}
	return &ctx, nil
}

func writeTopic(engine *storage.StorageEngine, idHash uint64, ctx *model.TopicSlot) error {
	data, err := json.Marshal(ctx)
	if err != nil {
		return core.NewError(core.ErrSerialization, "marshal topic", err)
	}
	_, err = engine.WriteRecord(storage.RecL2Topic, idHash, data)
	return err
}

func collectAllTopics(engine *storage.StorageEngine) []model.TopicSlot {
	var all []model.TopicSlot
	engine.IterIndex(func(idHash, _ uint64) bool {
		rt, data, err := engine.ReadRecord(idHash)
		if err != nil || rt != storage.RecL2Topic {
			return true
		}
		var ctx model.TopicSlot
		if json.Unmarshal(data, &ctx) == nil {
			all = append(all, ctx)
		}
		return true
	})
	return all
}

func filterByKeyword(all *[]model.TopicSlot, keyword *string) {
	if keyword == nil {
		return
	}
	kw := strings.ToLower(*keyword)
	filtered := make([]model.TopicSlot, 0, len(*all))
	for _, ctx := range *all {
		text := strings.ToLower(strings.Join(ctx.UserKeywords, " "))
		if strings.Contains(text, kw) {
			filtered = append(filtered, ctx)
		}
	}
	*all = filtered
}

func sortTopicsByUpdated(all []model.TopicSlot) {
	sort.Slice(all, func(i, j int) bool {
		return all[i].UpdatedAt > all[j].UpdatedAt
	})
}

func sortTopicsByCreated(all []model.TopicSlot) {
	sort.Slice(all, func(i, j int) bool {
		return all[i].CreatedAt < all[j].CreatedAt
	})
}

func toTopicSummary(ctx *model.TopicSlot) TopicSummary {
	turnCount := len(ctx.UserL4Refs) + len(ctx.AgentL4Refs)
	l3Count := len(ctx.UserL3Refs) + len(ctx.AgentL3Refs)
	return TopicSummary{
		ID:            hash.FormatHash(ctx.ID),
		Depth:         ctx.Depth,
		SceneID:       hash.FormatHash(ctx.SceneID),
		UserKeywords:  ctx.UserKeywords,
		AgentKeywords: ctx.AgentKeywords,
		FusedKeywords: ctx.FusedKeywords,
		FusedSummary:  ctx.FusedSummary,
		TurnCount:     turnCount,
		IsActive:      false,
		CreatedAt:     ctx.CreatedAt,
		L4Count:       turnCount,
		L3Count:       l3Count,
		UpdatedAt:     ctx.UpdatedAt,
	}
}

func toTopicDetail(ctx *model.TopicSlot) TopicDetail {
	return TopicDetail{
		ID:             hash.FormatHash(ctx.ID),
		ParentID:       formatOptUint64(ctx.ParentID),
		Depth:          ctx.Depth,
		SceneID:        hash.FormatHash(ctx.SceneID),
		UserKeywords:   ctx.UserKeywords,
		UserTimestamp:  ctx.UserTimestamp,
		AgentKeywords:  ctx.AgentKeywords,
		AgentTimestamp: ctx.AgentTimestamp,
		FusedKeywords:  ctx.FusedKeywords,
		FusedSummary:   ctx.FusedSummary,
		ChildrenIDs:    formatUint64Slice(ctx.ChildrenIDs),
		UserL4Refs:     formatUint64Slice(ctx.UserL4Refs),
		UserL3Refs:     formatUint64Slice(ctx.UserL3Refs),
		AgentL4Refs:    formatUint64Slice(ctx.AgentL4Refs),
		AgentL3Refs:    formatUint64Slice(ctx.AgentL3Refs),
		CreatedAt:      ctx.CreatedAt,
		UpdatedAt:      ctx.UpdatedAt,
	}
}

func applyL2Updates(ctx *model.TopicSlot, fields UpdateL2Fields) bool {
	changed := false
	if fields.UserKeywords != nil {
		ctx.UserKeywords = fields.UserKeywords
		changed = true
	}
	if fields.AgentKeywords != nil {
		ctx.AgentKeywords = fields.AgentKeywords
	}
	if fields.FusedSummary != nil {
		ctx.FusedSummary = fields.FusedSummary
		changed = true
	}
	if fields.L3Refs != nil {
		refs := parseUint64Slice(fields.L3Refs)
		sort.Slice(refs, func(i, j int) bool { return refs[i] < refs[j] })
		ctx.UserL3Refs = dedupUint64(refs)
		ctx.AgentL3Refs = nil
	}
	return changed
}

func reindexTopic(sparse *index.SparseIndex, ctx *model.TopicSlot) {
	sparse.RemoveDocument(ctx.ID)
	text := strings.Join(ctx.UserKeywords, " ")
	if ctx.FusedSummary != nil {
		text += " " + *ctx.FusedSummary
	}
	terms := index.Tokenize(text)
	sparse.AddDocument(ctx.ID, terms, uint32(len(terms)))
}

func deleteL1Nodes(engine *storage.StorageEngine, l1Reverse *L1ReverseIndex, ctxID uint64) {
	if l1Reverse == nil {
		return
	}
	ids := map[uint64]struct{}{ctxID: {}}
	nodeHashes := l1Reverse.FindAssociated(ids)
	for _, nodeHash := range nodeHashes {
		rt, _, err := engine.ReadRecord(nodeHash)
		if err == nil && rt == storage.RecL1SceneNode {
			engine.DeleteRecord(nodeHash)
			l1Reverse.RemoveNode(nodeHash)
		}
	}
}

func deleteL4Refs(engine *storage.StorageEngine, ctx *model.TopicSlot) {
	for _, ref := range ctx.UserL4Refs {
		engine.DeleteRecord(ref)
	}
	for _, ref := range ctx.AgentL4Refs {
		engine.DeleteRecord(ref)
	}
}

func parseMergeIDs(ids []string) ([]uint64, error) {
	hashes := make([]uint64, len(ids))
	for i, id := range ids {
		h, err := hash.ParseID(id)
		if err != nil {
			return nil, core.NewError(core.ErrInvalidQuery, "parse merge id", err)
		}
		if _, err2 := hash.ParseID(id); err2 != nil {
			return nil, err2
		}
		hashes[i] = h
	}
	return hashes, nil
}

func absorbSecondaries(
	engine *storage.StorageEngine,
	sparse *index.SparseIndex,
	primary *model.TopicSlot,
	mergeHashes []uint64,
	mergeIDs []string,
) {
	l4Set := toSet(primary.UserL4Refs)
	agL4Set := toSet(primary.AgentL4Refs)
	l3Set := toSet(primary.UserL3Refs)
	agL3Set := toSet(primary.AgentL3Refs)
	var summaries []string

	for i, secHash := range mergeHashes {
		sec, err := loadTopic(engine, secHash)
		if err != nil {
			continue
		}
		addAll(l4Set, sec.UserL4Refs)
		addAll(agL4Set, sec.AgentL4Refs)
		addAll(l3Set, sec.UserL3Refs)
		addAll(agL3Set, sec.AgentL3Refs)
		if sec.FusedSummary != nil {
			summaries = append(summaries, *sec.FusedSummary)
		}
		sparse.RemoveDocument(secHash)
		engine.DeleteRecord(secHash)
		_ = i
	}
	primary.UserL4Refs = sortedKeys(l4Set)
	primary.AgentL4Refs = sortedKeys(agL4Set)
	primary.UserL3Refs = sortedKeys(l3Set)
	primary.AgentL3Refs = sortedKeys(agL3Set)
	if len(summaries) > 0 {
		combined := ""
		if primary.FusedSummary != nil {
			combined = *primary.FusedSummary
		}
		for _, s := range summaries {
			if combined != "" {
				combined += " | "
			}
			combined += s
		}
		primary.FusedSummary = &combined
	}
}

func loadTopicsByIDs(engine *storage.StorageEngine, ids []uint64) []model.TopicSlot {
	var nodes []model.TopicSlot
	for _, id := range ids {
		rt, data, err := engine.ReadRecord(id)
		if err != nil || rt != storage.RecL2Topic {
			continue
		}
		var ctx model.TopicSlot
		if json.Unmarshal(data, &ctx) == nil {
			nodes = append(nodes, ctx)
		}
	}
	return nodes
}

func buildSceneTree(sceneID uint64, nodes []model.TopicSlot) *SceneTreeResult {
	totalTurns := uint32(len(nodes))
	var depthDist [4]uint32
	var edges [][2]string
	details := make([]TopicDetail, len(nodes))
	for i, ctx := range nodes {
		depthIdx := int(ctx.Depth) - 1
		if depthIdx < 0 {
			depthIdx = 0
		}
		if depthIdx > 3 {
			depthIdx = 3
		}
		depthDist[depthIdx]++
		if ctx.ParentID != nil {
			edges = append(edges, [2]string{hash.FormatHash(*ctx.ParentID), hash.FormatHash(ctx.ID)})
		}
		for _, childID := range ctx.ChildrenIDs {
			edges = append(edges, [2]string{hash.FormatHash(ctx.ID), hash.FormatHash(childID)})
		}
		details[i] = toTopicDetail(&nodes[i])
	}
	return &SceneTreeResult{
		SceneID:           hash.FormatHash(sceneID),
		TotalTurns:        totalTurns,
		DepthDistribution: depthDist,
		Nodes:             details,
		Edges:             edges,
	}
}

func emptySceneTree(sceneID uint64) *SceneTreeResult {
	return &SceneTreeResult{
		SceneID: hash.FormatHash(sceneID),
		Nodes:   []TopicDetail{},
		Edges:   [][2]string{},
	}
}

// --- pagination & format helpers ---

func paginationParams(page, pageSize int) (skip, take int) {
	if page < 1 {
		page = 1
	}
	if pageSize < 1 {
		pageSize = 20
	}
	return (page - 1) * pageSize, pageSize
}

func formatOptUint64(v *uint64) *string {
	if v == nil {
		return nil
	}
	s := hash.FormatHash(*v)
	return &s
}

func formatUint64Slice(ids []uint64) []string {
	out := make([]string, len(ids))
	for i, id := range ids {
		out[i] = hash.FormatHash(id)
	}
	return out
}

func parseUint64Slice(ids []string) []uint64 {
	out := make([]uint64, 0, len(ids))
	for _, s := range ids {
		h, err := hash.ParseID(s)
		if err == nil {
			out = append(out, h)
		}
	}
	return out
}

func toSet(ids []uint64) map[uint64]struct{} {
	s := make(map[uint64]struct{}, len(ids))
	for _, id := range ids {
		s[id] = struct{}{}
	}
	return s
}

func addAll(s map[uint64]struct{}, ids []uint64) {
	for _, id := range ids {
		s[id] = struct{}{}
	}
}

func sortedKeys(s map[uint64]struct{}) []uint64 {
	out := make([]uint64, 0, len(s))
	for k := range s {
		out = append(out, k)
	}
	sort.Slice(out, func(i, j int) bool { return out[i] < out[j] })
	return out
}

func dedupUint64(ids []uint64) []uint64 {
	if len(ids) == 0 {
		return ids
	}
	out := []uint64{ids[0]}
	for i := 1; i < len(ids); i++ {
		if ids[i] != ids[i-1] {
			out = append(out, ids[i])
		}
	}
	return out
}
